server/src/compat/samplers/retained_batch_sampler.rs

```diff
@@ -0,0 +1,63 @@
+use crate::compat::binary_schema::BinarySchema;
+use crate::compat::schema_sampler::BinarySchemaSampler;
+use crate::compat::snapshots::retained_batch_snapshot::RetainedMessageBatchSnapshot;
+use crate::streaming::utils::file;
+use async_trait::async_trait;
+use bytes::{BufMut, Bytes};
+use iggy::error::IggyError;
+use tokio::io::AsyncReadExt;
+
+pub struct RetainedMessageBatchSampler {
+    pub segment_start_offset: u64,
+    pub log_path: String,
+    pub index_path: String,
+}
+
+impl RetainedMessageBatchSampler {
+    pub fn new(
+        segment_start_offset: u64,
+        log_path: String,
+        index_path: String,
+    ) -> RetainedMessageBatchSampler {
+        RetainedMessageBatchSampler {
+            segment_start_offset,
+            log_path,
+            index_path,
+        }
+    }
+}
+
+unsafe impl Send for RetainedMessageBatchSampler {}
+unsafe impl Sync for RetainedMessageBatchSampler {}
+
+#[async_trait]
+impl BinarySchemaSampler for RetainedMessageBatchSampler {
+    async fn try_sample(&self) -> Result<BinarySchema, IggyError> {
+        let mut index_file = file::open(&self.index_path).await?;
+        let mut log_file = file::open(&self.log_path).await?;
+        let log_file_size = log_file.metadata().await?.len();
+
+        if log_file_size == 0 {
+            return Ok(BinarySchema::RetainedMessageBatchSchema);
+        }
+
+        let _ = index_file.read_u32_le().await?;
+        let _ = index_file.read_u32_le().await?;
+        let second_index_offset = index_file.read_u32_le().await;
```

Comment:
Once we'll store just an index value, this offset and the one above will no longer be needed.